website/blog/2023-06-14-package-exports-support.md

```diff
@@ -0,0 +1,235 @@
+:::
+
+## For app developers
+
+Package Exports can be enabled today, experimentally.
+
+- Imports against packages that depend on Package Exports features (such as [**Firebase**](https://www.npmjs.com/package/firebase) and [**Storybook**](https://www.npmjs.com/search?q=%40storybook)) should now work as designed.
+- React Native for Web projects using Metro will now be able to use the `"browser"` conditional export, removing the need for workarounds.
+
+Enabling Package Exports brings a few [edge-case breaking changes](#breaking-changes) that may affect specific projects, and which you can [test today](#validating-changes-in-your-project).
+
+**In a future React Native release, Package Exports will be enabled by default**. In a chicken-and-egg situation, React Native apps were previously a holdout for some packages to migrate to `"exports"` — or used our `"react-native"` root field escape hatch. Supporting these features in Metro will allow the ecosystem to move forward.
+
+### Enabling Package Exports (experimental)
+
+Package Exports can be enabled in your app's [**metro.config.js**](https://github.com/facebook/react-native/blob/0.72-stable/packages/react-native/template/metro.config.js) file via the [`resolver.unstable_enablePackageExports`](https://facebook.github.io/metro/docs/configuration/#unstable_enablepackageexports-experimental) option.
+
+```js
+const config = {
+  // ...
+  resolver: {
+    unstable_enablePackageExports: true,
+  },
+};
+```
+
+Metro exposes two further resolver options which configure how conditional exports behave:
+
+- [`unstable_conditionNames`](https://facebook.github.io/metro/docs/configuration/#unstable_conditionnames-experimental) — The set of [condition names](https://nodejs.org/docs/latest-v19.x/api/packages.html#community-conditions-definitions) to assert when resolving conditional exports. By default, we match `['require', 'import', 'react-native']`.
+- [`unstable_conditionsByPlatform`](https://facebook.github.io/metro/docs/configuration/#unstable_conditionsbyplatform-experimental) — The additional condition names to assert when resolving for a given platform target. By default, this matches `'browser'` when the platform is `'web'`.
+
+:::tip
+**Remember to use the React Native [Jest preset](https://github.com/facebook/react-native/blob/main/template/jest.config.js#L2)!** Jest includes support for Package Exports by default. In tests, you can override which `customExportConditions` are resolved using the [`testEnvironmentOptions`](https://jestjs.io/docs/configuration#testenvironmentoptions-object) option.
```

Comment:
We are doing something similar in the RN 0.73 template `tsconfig`, but I did actually walk back the change I made in RN 0.72 to move to `moduleResolution: 'nodenext'`, because it didn't make sense to resolve by default when Metro wasn't yet, and the version of TypeScript shipped with 0.72 had some limitations.

So, TypeScript users may need to change their `moduleResolution` setting to enable package exports as well.

Reply:
Good point! I've extended this note like so: 

<img width="814" alt="image" src="https://github.com/facebook/react-native-website/assets/2547783/f1f49120-3e49-451f-b82a-60b5b5138378">

Should only affect edge cases, and yeah we'll have this aligned by default for RN 0.73.

Reply:
`moduleResolution: 'bundler'` does have a caveat that it is not supported by TypeScript < 5, so the version of TypeScript we ship out of the box does not support it.

The closest thing supported in older versions to enable following exports is `moduleResolution: "nodenext"`, but it has some restrictions I'm not sure is consistent with Metro.